Add next-gen future documentation (#761)
slackapi/bolt-python: docs/_future/getting_started_future.md

---
title: Getting started
order: 1
slug: getting-started-future
lang: en
layout: tutorial
permalink: /tutorial/getting-started-future
---

## Getting started <span class="label-beta">BETA</span>

<div class="section-content">
This guide will cover how to get started with your next-gen platform using Bolt for Python, by setting up the Slack CLI and installing the required dependencies.

Find out about the next-generation platform on Slack's <a href="https://api.slack.com/future/intro" target="_blank">official introduction page.</a>
</div>

---

### Limitations

> Bolt for Python supports app development using next-gen platform features like Functions, Workflows and tools such as the Slack CLI alongside all current generally available Slack Platform features.

Reviewer: Let's link to the Functions and Workflows pages you added in this PR in this line.

#### We do not yet support

- Deployment to secure and managed Slack infrastructure.
- Datastores API <a href="https://api.slack.com/future/datastores" target="_blank">Datastores</a> functionality.

> > If you'd like to deploy your app with Slack infrastructure, consider building your next-generation application with the Deno Slack API. You can get started with that <a href="https://api.slack.com/future/get-started" target="_blank">here</a>.

Reviewer 2: ```suggestion
> 💡 If you'd like to deploy your app with Slack infrastructure, consider building your next-generation application with the Deno Slack API. You can get started with that <a href="https://api.slack.com/future/get-started" target="_blank">here</a>.
```

---

### Setting up {#setting-up}

#### Slack CLI {#setting-up-cli}

To build a next-generation app with Bolt for Python, you'll need to get the Slack CLI.

Install the Slack CLI by following this <a href="https://api.slack.com/future/quickstart" target="_blank">Quickstart</a>. Since we won't be using Deno to build our next-generation app, you can skip any instructions related to installing Deno or creating an app using a Deno template. Once you've logged into the CLI using `slack login` and verified your login using `slack auth list`, you can proceed with the instructions in this guide.

#### Dependencies {#setting-up-dependencies}

> Once the CLI is set up, make sure your machine has the most recent version of <a href="https://www.python.org/" target="_blank">Python</a> installed. You can install Python through a package manager (such as <a href="https://brew.sh/" target="_blank">Homebrew</a> for Mac) or directly from the <a href="https://www.python.org/downloads/" target="_blank">website</a>.

Reviewer 3: ```suggestion
Once the CLI is set up, make sure your machine has the most recent version of <a href="https://www.python.org/" target="_blank">Python</a> installed. You can install Python through a package manager (such as <a href="https://brew.sh/" target="_blank">Homebrew</a> for macOS) or directly from the <a href="https://www.python.org/downloads/" target="_blank">website</a>.
```

Once the app is successfully created, you should see a message like this:

```text
✨ my-app successfully created

🧭 Explore your project's README.md for documentation and code samples, and at any time run slack help to display a list of available commands

🧑‍🚀 Follow the steps below to try out your new project

1️⃣  Change into your project directory with: cd my-app

2️⃣  Develop locally and see changes in real-time with: slack run

3️⃣  When you're ready to deploy for production use: slack deploy

🔔 If you leave the workspace, you won’t be able to manage any apps you’ve deployed to it. Apps you deploy will belong to the workspace even if you leave the workspace
```

---

### Set up your trigger {#setup-trigger}

> As mentioned, this app comes with pre-existing functionality - it uses Functions, Workflows and a Trigger set up to kick things off. Let's run a command to initialize that Trigger via the CLI.

Reviewer: Let's link the capital letter terms to the relevant document sections!

I think we should also explain _why_ the user would want to initialize a Trigger, otherwise we're not teaching developers what this concept is or why they should care about them. Maybe something like:

```suggestion
As mentioned, this app comes with pre-existing functionality - it uses Functions, Workflows and a Link Trigger (link to link trigger docs: https://api.slack.com/future/triggers/link) that will allow users in Slack to initiate the functionality provided by the app. Let's run a command to initialize that Link Trigger via the CLI.
```

First, make sure you're in the project directory in your command line: `cd my-app`

Then, run the following command to create a Trigger:

```bash
slack triggers create --trigger-def "triggers/sample-trigger.json"      
```

> The above command will create a trigger for the selected workspace. Make sure to select the workspace you want. Once the trigger is successfully created, you should see an output like this:

Reviewer: ```suggestion
The above command will create a [Link Trigger](https://api.slack.com/future/triggers/link) for the selected workspace. Make sure to select the workspace you want. Once the trigger is successfully created, you should see an output like this:
```

> The above command will create a trigger for the selected workspace. Make sure to select the workspace you want and that it is appended by `(dev)`. This will create a dev instance of your app. Once the trigger is successfully created, you should see an output like this:

Reviewer 4: This selector feature may change in the future. Thoughts on just saying "Select the workspace you want to trigger your workflow from. Once the trigger is successfully created, you should see an output like this:"?

```suggestion
The above command will create a trigger for the selected workspace. Make sure to select the workspace you want and that it is appended by `(dev)`. This will create a dev instance of your app. Once the trigger is successfully created, you should see an output like this:
```

```bash
⚡ Trigger created
   Trigger ID:   [ID]
   Trigger Type: shortcut
   Trigger Name: Sample Trigger
   URL: https://slack.com/shortcuts/[ID]/[Some ID]
```

> The provided URL will be what you use to run your Trigger. Copy this URL and save it somewhere; you'll need it for later.

Reviewer: ```suggestion
The provided URL can be pasted into Slack; Slack will unfurl it into a button that users can interact with to initiate your app's functionality! Copy this URL and save it somewhere; you'll need it for later.
```

---

### Run your app {#run-your-app}

Now that your app and Trigger are successfully created, let's try running it!

```bash
# install the required project dependencies
pip install -r requirements.txt

# start a local development server
slack run
```

> Executing `pip install -r requirements.txt` installs all the project requirements to your machine.

Reviewer: We may want to link to any existing documentation we have around setting up a virtual env, as that is the recommended way of working in Python projects.

Author: Yeah I've been thinking about doing this in our documentation, currently this is how we do it in the rest of our docs

Reviewer: Let's chalk it up as a separate issue so we don't forget, and if/when we get to it we can update all the documentation (new platform or otherwise) to link to it where it makes sense.

Author: I created issue #777 to keep track of this

Executing `slack run` starts a local development server, syncing changes to your workspace's development version of your app.

You'll be prompted to select a workspace to install the app to&mdash;select the development instance of your workspace (you'll know it's the development version because the name has the string `(dev)` appended).

> 💡 If you don't see the workspace you'd like to use in the list, you can `CTRL + C` out of the `slack run` command and run `slack auth login`. This will allow you to authenticate in your desired workspace to have it show up in the list for `slack run`.

> Once the app is successfully ran, you'll see output in your Terminal to indicate it's running, similar to what you would see with any other Bolt for Python app. You can search for the `⚡️ Bolt app is running! ⚡️` message to make sure that your app has successfully started up.

Reviewer 4: ```suggestion
You'll see output in your Terminal to indicate it's running, similar to what you would see with any other Bolt for Python app. You can search for the `⚡️ Bolt app is running! ⚡️` message to make sure that your app has successfully started up.
```
